pipeline_server/src/runner.rs

```diff
@@ -0,0 +1,479 @@
+
+        match Self::wait_for_startup(&self.config.log_file_path(pipeline_id)).await {
+            Ok(port) => {
+                // Store pipeline in the database.
+                if let Err(e) = self
+                    .db
+                    .lock()
+                    .await
+                    .new_pipeline(
+                        pipeline_id,
+                        request.project_id,
+                        request.project_version,
+                        port,
+                    )
+                    .await
+                {
+                    let _ = pipeline_process.kill().await;
+                    return Err(e);
+                };
+                let json_string =
+                    serde_json::to_string(&NewPipelineResponse { pipeline_id, port }).unwrap();
+
+                // Create Prometheus config file for the pipeline.
+                // The Prometheus server should pick up this file automatically.
+                self.create_prometheus_config(
+                    &project_descr.name,
+                    request.project_id,
+                    pipeline_id,
+                    port,
+                )
+                .await
+                // Don't abandon pipeline, just log the error.
+                .unwrap_or_else(|e| {
+                    error!(
+                        "Failed to create Prometheus config file for pipeline '{pipeline_id}': {e}"
+                    );
+                });
+
+                Ok(HttpResponse::Ok()
+                    .content_type(mime::APPLICATION_JSON)
+                    .body(json_string))
+            }
+            Err(e) => {
+                let _ = pipeline_process.kill().await;
+                Err(e)
+            }
+        }
+    }
+
+    async fn start(
+        &self,
+        db: &ProjectDB,
+        request: &NewPipelineRequest,
+        config_yaml: &str,
+        pipeline_id: PipelineId,
+    ) -> AnyResult<Child> {
+        // Create pipeline directory (delete old directory if exists); write metadata
+        // and config files to it.
+        let pipeline_dir = self.config.pipeline_dir(pipeline_id);
+        create_dir_all(&pipeline_dir).await?;
+
+        // let config_yaml = self.create_topics(config_yaml).await?;
+
+        let config_file_path = self.config.config_file_path(pipeline_id);
+        fs::write(&config_file_path, config_yaml).await?;
+
+        let (_version, code) = db.project_code(request.project_id).await?;
+
+        let metadata = PipelineMetadata {
+            project_id: request.project_id,
+            version: request.project_version,
+            code,
+        };
+        let metadata_file_path = self.config.metadata_file_path(pipeline_id);
+        fs::write(
+            &metadata_file_path,
+            serde_json::to_string(&metadata).unwrap(),
+        )
+        .await?;
+
+        let log_file_path = self.config.log_file_path(pipeline_id);
+        let log_file = File::create(&log_file_path).await?;
```

Comment:
do you need to?

Reply:
yeah, the `stderr` method takes a file handle, not just filename.